[MSE] media element will error appending the same media segment twice

jyavenard opened the pull request:
#### 7d2a7b505c37c4bdb7863691dd45ac113d3265fc
<pre>
[MSE] media element will error appending the same media segment twice
<a href="https://bugs.webkit.org/show_bug.cgi?id=271685">https://bugs.webkit.org/show_bug.cgi?id=271685</a>
<a href="https://rdar.apple.com/125386530">rdar://125386530</a>

Reviewed by Youenn Fablet and Eric Carlson.

The test file test-fragmented.mp4 was actually invalid and had its tfhd&apos;s
base-data-offset-present flag set to true, which is against the spec as per
<a href="https://www.w3.org/TR/mse-byte-stream-format-isobmff/#movie-fragment-relative-addressing.">https://www.w3.org/TR/mse-byte-stream-format-isobmff/#movie-fragment-relative-addressing.</a>
This prevented the media segment from being added out of order (or twice).
Re-mux the file with the MP4Box utility from the GPAC suite (`MP4Box -flat -dash 1000 -profile onDemand`)

Add test that expands on webkit.org/b/270618.
Updates existing tests due to slight changes of frames timing.

* LayoutTests/http/tests/webcodecs/h264-reordering.html: Amend offsets to new mp4.
* LayoutTests/media/media-source/content/test-fragmented-manifest.json:
* LayoutTests/media/media-source/content/test-fragmented.mp4:
* LayoutTests/media/media-source/media-managedmse-append-expected.txt: Added.
* LayoutTests/media/media-source/media-managedmse-append.html: Added.
* LayoutTests/media/media-source/content/test-fragmented-manifest.json:
* LayoutTests/media/media-source/media-managedmse-resume-after-stall-expected.txt:
* LayoutTests/media/media-source/media-managedmse-resume-after-stall.html: video and audio track do not exactly end at 3s (audio is slightly behind).
Adjust test to reflect the new buffered range.
* LayoutTests/media/media-source/media-managedmse-stall-endofstream-expected.txt:
* LayoutTests/media/media-source/media-managedmse-stall-endofstream.html:
* LayoutTests/media/media-source/media-source-trackid-change-expected.txt:
* LayoutTests/media/media-source/media-source-trackid-change.html: update test to adjust offsets into binary mp4 file. Add comments referring to what those offsets are.
* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::monitorSourceBuffers): simplify code a bit.

Canonical link: <a href="https://commits.webkit.org/276821@main">https://commits.webkit.org/276821@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/62d843d675fd9d8399bb29dc0148f800f877b715

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45761 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24888 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48342 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48431 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41796 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48067 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29170 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22286 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37466 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46339 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21968 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39481 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18623 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19368 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40574 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3805 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42072 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40918 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50186 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20752 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17262 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44581 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22057 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/39637 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43440 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10166 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22416 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21746 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf reviewed (approved):
Would be nice to write the test as WPT...

eric-carlson reviewed (approved):
Nit: s/increased passed the/increased past the/

jyavenard committed:
[MSE] media element will error appending the same media segment twice
https://bugs.webkit.org/show_bug.cgi?id=271685
rdar://125386530

Reviewed by Youenn Fablet and Eric Carlson.

The test file test-fragmented.mp4 was actually invalid and had its tfhd's
base-data-offset-present flag set to true, which is against the spec as per
https://www.w3.org/TR/mse-byte-stream-format-isobmff/#movie-fragment-relative-addressing.
This prevented the media segment from being added out of order (or twice).
Re-mux the file with the MP4Box utility from the GPAC suite (`MP4Box -flat -dash 1000 -profile onDemand`)

Add test that expands on webkit.org/b/270618.
Updates existing tests due to slight changes of frames timing.

* LayoutTests/http/tests/webcodecs/h264-reordering.html: Amend offsets to new mp4.
* LayoutTests/media/media-source/content/test-fragmented-manifest.json:
* LayoutTests/media/media-source/content/test-fragmented.mp4:
* LayoutTests/media/media-source/media-managedmse-append-expected.txt: Added.
* LayoutTests/media/media-source/media-managedmse-append.html: Added.
* LayoutTests/media/media-source/content/test-fragmented-manifest.json:
* LayoutTests/media/media-source/media-managedmse-resume-after-stall-expected.txt:
* LayoutTests/media/media-source/media-managedmse-resume-after-stall.html: video and audio track do not exactly end at 3s (audio is slightly behind).
Adjust test to reflect the new buffered range.
* LayoutTests/media/media-source/media-managedmse-stall-endofstream-expected.txt:
* LayoutTests/media/media-source/media-managedmse-stall-endofstream.html:
* LayoutTests/media/media-source/media-source-trackid-change-expected.txt:
* LayoutTests/media/media-source/media-source-trackid-change.html: update test to adjust offsets into binary mp4 file. Add comments referring to what those offsets are.
* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::monitorSourceBuffers): simplify code a bit.

Canonical link: https://commits.webkit.org/276821@main
